Add particle unique ID (#351)
gamer-project/gamer: doc/wiki/Runtime-Parameters-related/[Runtime-Parameters]-Particles.md

[[Setting IC from Files &#8212; Particles | Initial Conditions#IC-File-Particles]].
    * **Restriction:**

<a name="PAR_IC_PUID"></a>
* #### `PAR_IC_PUID` &ensp; (0=off, 1=on) &ensp; [0]
    * **Description:**
Check whether all existing particles have UIDs assigned when [PAR_INIT](#PAR_INIT)=3 is used. Note that when the particle initial condition file `PAR_IC`
does not include the particle UID data,

> one should turn this off and the new particle UIDs will be assigned internally.

Reviewer: ```suggestion
Flag indicating whether the particle initial condition file `PAR_IC` contains UID information.
If this option is disabled, new particle UIDs will be assigned automatically.
```

## Remarks

### Particle UID (`PUID`)
Valid particle UIDs should lie within `[1, number of particles]`. If new particles are created
during the simulation, please assign their PUIDs to `PUID_TBA`, and then call `Par_SetParUID()`
to assign PUIDs after all particle creation within the same routine is complete.

The particle UID is assigned only in the following two situations:
1. Initialization

   Particle UIDs are assigned after `Par_Init_ByFunction_Ptr()` or `Par_Init_ByFile()`.
   First, we collect the particle UIDs from all ranks into a single array.
   Second, we assign the particle UIDs according to the array index plus one.
   Finally, send the particle UIDs back to all ranks.
2. During simulation

   Particle UIDs should be assigned _after_ all the particle creation within a
   single routine is complete (e.g. star formation).
   First, we collect the PUIDs and positions of new particles from all ranks
   into a single array.
   Second, we sort the particles by their positions.
   Third, assign the particle UIDs according to the array index plus `NextPUID`.
   Finally, send the new particle UIDs back to all ranks.

>    Example: `src/StarFormation/SF_CreateStar.cpp` and `src/StarFormation/SF_CreateStar_AGORA.cpp`

Reviewer 2: ```suggestion
Valid particle UIDs should lie within `[1, number of particles]`.

The particle UID assignment occurs during either initialization or on-the-fly creation 
in simulation (e.g., star formation). The particle UIDs can be inherited from `PAR_IC` 
(when `PAR_INIT=3` and `PAR_IC_PUID` is enabled) or from a restart file. Except for 
these, the particle UIDs should be assigned internally via `Par_SetParUID()` in all 
other cases (e.g., `PAR_INIT=1`).

To ensure uniqueness and reproducibility, all particle UIDs should be assigned by 
`Par_SetParUID()` globally, rather than decided by individual MPI ranks. The particle 
UID must be first marked as `PUID_TBA` when the particle is created (e.g., in 
`Par_Init_ByFunction_Ptr()`, `Par_Init_ByFile()`, 
`src/StarFormation/SF_CreateStar_AGORA.cpp`) before invoking `Par_SetParUID()` (e.g., at 
the end of `Init_GAMER()` and `SF_CreateStar()`). And `Par_SetParUID()` should be 
invoked _after_ all the particle creation within a single routine is complete.

The assignment procedure of `Par_SetParUID()` follows:

   First, we collect the PUIDs and positions of new particles from all ranks
   into a single array.
   Second, we sort the particles by their positions.
   Third, assign the particle UIDs according to the array index plus `NextPUID`.
   Finally, send the new particle UIDs back to all ranks.
```

## Remarks

### Particle UID
The valid particle UID should be in `[1, number of particles]`. If there is
a new particle created during the simulation, please assign the particle
UID to `-1` and then call `Par_SetUID(false)` to assign UID after all creations
are done in a single routine.

The particle UID is assigned in only the following two situations:
1. Initialization

   The particle UID is assigned after `Par_Init_ByFunction_Ptr()` or `Par_Init_ByFile()`.
   First, we collect the UID of particles from all ranks to a single array.
   Second, we assign the particle UID by the array index plus one.
   Finally, send the particle UIDs back to all ranks.
2. During simulation

   The particle UID should be assigned _after_ all the creations are done
   in a single routine (e.g. star formation).
   First, we collect the UID and position of new particles from all ranks
   to a single array.
   Second, we sort the particles by their position.
   Third, assign the particle UID by the array index plus `NextUID`.

>    Finally, send the new particle UIDs back to all ranks.

Reviewer: ```suggestion
1. Initialization

   Particle UIDs are assigned after `Par_Init_ByFunction_Ptr()` or `Par_Init_ByFile()`.
   First, we collect the particle UIDs from all ranks into a single array.
   Second, we assign the particle UIDs according to the array index plus one.
   Finally, send the particle UIDs back to all ranks.
2. During simulation

   Particle UIDs should be assigned _after_ all the particle creation within a
   single routine is complete (e.g. star formation).
   First, we collect the UIDs and positions of new particles from all ranks
   into a single array.
   Second, we sort the particles by their positions.
   Third, assign the particle UIDs according to the array index plus `NextUID`.
   Finally, send the new particle UIDs back to all ranks.
```

   Example: `src/StarFormation/SF_CreateStar.cpp` and `src/StarFormation/SF_CreateStar_AGORA.cpp`

> [!CAUTION]

> > If the particle positions are exactly the same (which should unlikely happen), the UID is decremented by the following order:

Reviewer: ```suggestion
> If the particle positions are exactly the same (which should unlikely happen), the UID is determined by the following order:
```

> 1. The rank number (`MPI_Rank`) from 0 to `MPI_NRank-1`.

> > 2. The index of particle array in each rank.

Reviewer: I'm not sure whether we should specify this, since the results may depend on how the sorting is implemented. What do you think?
